[8.0] Add arguments to DIRAC.initialize (#6335)
DIRACGrid/DIRAC: docs/source/UserGuide/Tutorials/UsingDIRACFromPython/index.rst

.. _usingDIRACFromPython:

===============================

> Using DIRAC Clients From Python

Author: There are a lot of users that write python scripts against DIRAC for whom the developer documentation is excessive. I think there is a place for a tutorial that only documents clients without details about configuration/services/agents/testing.

===============================

.. toctree::
   :maxdepth: 2

Overview
--------

> For most tasks it is sufficient to use the DIRAC from a terminal via the commands that are prefixed with ``dirac-`` or via the WebApp.

Reviewer: ```suggestion
For most tasks it is sufficient to use DIRAC from a terminal via the commands that are prefixed with ``dirac-`` or via the WebApp.
```

This function can be used as follows:

.. code-block:: python

    from DIRAC.Core.Utilities.Network import getIPsForHostName

    for domain in ["diracgrid.org", "fake-domain.invalid"]:
        ret = getIPsForHostName(domain)
        if not ret["OK"]:
            raise RuntimeError(f"Failed to find IPs for {domain}.org with error: {ret['Message']}")
        ips = ret["Value"]
        print(f"{domain} is running at {ips}")

This will print:

.. code-block:: python

    diracgrid.org is running at ['134.158.16.209']
    Traceback (most recent call last):
    File "return-values-example.py", line 6, in <module>
        raise RuntimeError(f"Failed to find IPs for {domain}.org with error: {ret['Message']}")
    RuntimeError: Failed to find IPs for fake-domain.invalid.org with error: Can't get info for host fake-domain.invalid: [Errno 8] nodename nor servname provided, or not known

Often when writing user scripts it is useful to immediately raise an exception when an error happens and otherwise just return the value.

> This is makes DIRAC behave more similarly to other Python functions and can be achieved using the helper function :py:func:`~DIRAC.Core.Utilities.ReturnValues.returnValueOrRaise`.

Reviewer: ```suggestion
This makes DIRAC behave more similarly to other Python functions and can be achieved using the helper function :py:func:`~DIRAC.Core.Utilities.ReturnValues.returnValueOrRaise`.
```

Rewriting our previous example for looking up IP addressed, this would be:

.. code-block:: python

    from DIRAC.Core.Utilities.ReturnValues import returnValueOrRaise
    from DIRAC.Core.Utilities.Network import getIPsForHostName

    for domain in ["diracgrid.org", "fake-domain.invalid"]:
        ips = returnValueOrRaise(getIPsForHostName(domain))
        print(f"{domain} is running at {ips}")

.. _rpcCalls:

Remote Procedure Calls
^^^^^^^^^^^^^^^^^^^^^^

> Remote Procedure Calls (RPC) describes the process of sending a request to another computer (i.e. DIRAC servers) that triggers them to process something.

Reviewer: ```suggestion
Remote Procedure Calls (RPC) describe the process of sending a request to other computers (i.e. DIRAC servers) that triggers them to process something.
```

The result of this remotely ran function is then sent back to the origin of the call (i.e. your computer).

In DIRAC this is done automatically when using the ``Client`` classes where calling a method ``MyClient().something(123)`` triggers the server to execute a function named ``export_something(123)``.
See :ref:`usingAClient` for an example of using a client.

Initializing DIRAC
------------------

Before using Python to contact a service DIRAC's global state needs to be initialized.
Most importantly, this process contacts the Configuration Service and starts a refresher thread.
When writing a Python script this should be done using the :py:func:`DIRAC.initialize` function.

.. code-block:: python

   import DIRAC
   DIRAC.initialize()

> **Note** Currently is it essential to call :py:func:`~DIRAC.initialize` before importing clients.

Author: Done

> **Note** Currently it is essential to call :py:func:`~DIRAC.initialize` before importing clients.

Reviewer: ```suggestion
.. note:: Currently it is essential to call :py:func:`~DIRAC.initialize` before importing clients.
```
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-note

> **Note** If you're writing a ``dirac-`` command initialization should be handled differently. See :ref:`developingCommands`.

Reviewer: ```suggestion
.. note:: If you're writing a ``dirac-`` command initialization should be handled differently. See :ref:`developingCommands`.
```
